### PR TITLE
Improvements to inferConstRefs

### PR DESCRIPTION
**Pull request by benharsh**
RVF should only forward references to const things
I made the naive assumption that RVF should be allowed to forward any
argument with INTENT_CONST_REF. This shouldn't be allowed if the
argument might alias another non-const-ref argument to the on function.

The solution, I believe, is to improve inferConstRefs such that it can
correctly apply FLAG_REF_TO_CONST. RVF will then look for this flag when
deciding whether or not to forward an argument.

The first part of making such a determination is to identify Symbols
with QUAL_VAL and determine if they could have QUAL_CONST_VAL instead.
This is currently done in a very conservative way, only accepting
something as const-val if:
  1) It's passed to a formal with const-ref
  2) It's used in an ADDR_OF where the LHS is a const-ref
  3) It's only used on the LHS of a MOVE/ASSIGN once
In all other cases, I take the safe approach and assume the symbol is
not const.

Now that a number of references and vals have constness info, I can go
about inferring if something is a reference to a const. Here are the
acceptable conditions:
  1) If the symbol is a formal, it is ref-to-const if all corresponding
  actuals are also ref-to-const or const-val.
  2) If on the LHS of a move/assign...
    2a) and the RHS is a reference, the RHS must also be ref-to-const
    2b) and the RHS is a val, it must be a const-val
  3) It's used in one of a number of primitives that do not change the
  const-ness or address of the reference, as defined by the helper
  function isSafeRefPrimitive.

I also adjusted the re-entrant functions to only change properties on
symbols iff they were the first to be called on a symbol.

RVF will now also forward const-refs to classes. Otherwise this would generate a wide-reference to a wide class, and cause communication just to get to the wide class pointer.

A number of tests' comm counts have regressed now that RVF is correctly forwarding fewer things.

I have also added some RVF tests under optimizations/remoteValueForwarding/bharshbarg concerning correctness and future hopes for RVF's effectiveness.

### CODE REVIEW

**Comment by benharsh** [comment]
Testing:
- [x] full local
- [x] full no-local
- [x] full gasnet
- [x] dist-cyclic
- [x] dist-replicated

File: compiler/optimizations/remoteValueForwarding.cpp [hunk 1]
```diff
@@ -358,14 +358,24 @@ static bool canForwardValue(Map<Symbol*, Vec<SymExpr*>*>& defMap,
...
+    if (info && info->usesDotLocale) {
+      retval = false;
+    } else if (arg->intent == INTENT_CONST_REF) {
+      if (isClass(arg->getValType())) {
+        // When passing a reference to a class across an on-statement, it will
+        // normally be generated as a wide-reference to a wide-class. When the
+        // reference is a const-ref, this introduces unnecessary communication
+        // to simply get to the wide class pointer. Because the reference is
+        // never written to, we can simply RVF the class pointer.
+        retval = true;
```
Comment (mppf) [inline compiler/optimizations/remoteValueForwarding.cpp]: Doesn't this run the same risk as for records; where another task might modify a reference actual before the on-statement body uses it?
It might be that we know for other reasons that this is O.K... but it's not obvious to me at first glance.

File: test/optimizations/remoteValueForwarding/bharshbarg/alias.chpl [hunk 2]
```diff
@@ -0,0 +1,31 @@
+
+var g : int = 0;
+
+var gr : R;
+
+record R {
+  var x : int;
+}
```
Comment (mppf) [inline test/optimizations/remoteValueForwarding/bharshbarg/alias.chpl]: How about making a corresponding version using a class and a global reference to to a class instance, to test the code I was just pointing out?

e.g.

class C { ... }
var instance = new C(...);
ref globalRef = instance;

... testClass(globalRef) ...

proc testClass(const ref globalRef) {
 ...
 globalRef.x += 1;
 ...
}

File: compiler/optimizations/remoteValueForwarding.cpp (hunk 1 above)
Comment (mppf) [inline compiler/optimizations/remoteValueForwarding.cpp]: Oh, right. The language prohibits references from being modified. So it would have to be some compiler-introduced code. Should be OK.

File: compiler/optimizations/inferConstRefs.cpp [hunk 3]
```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
...
+// 2) QUAL_CONST_REF / INTENT_CONST_REF
+// 3) FLAG_REF_TO_CONST
 //
 void inferConstRefs() {
-  // Build a map from Symbols to RefInfo. Like buildDefUseMaps, except we don't
-  // care about the distinction between a def and a use. We just want all
-  // SymExprs for a Symbol.
+  // Build a map from Symbols to ConstInfo. This is somewhat like
+  // buildDefUseMaps, except we don't want to put defs and uses in different
+  // lists (for simplicity).
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: Did you try using for_SymbolSymExprs ?

File: compiler/optimizations/inferConstRefs.cpp [hunk 4]
```diff
@@ -23,62 +23,109 @@
...
-    Symbol*              sym;
-    std::deque<SymExpr*> todo;
-    RefInfo(Symbol*);
+    bool                  finalizedConstness;
+    bool                  finalizedRefToConst;
+    bool                  alreadyCalled;
+    Symbol*               sym;
+    SymExpr*              fnUses;
+
+    // TODO: Should we use the linked-list embedded in each Symbol?
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: I'd recommend trying that in stages - first just changing the loop building up ConstInfo to no longer use gSymExprs and then moving that list traversal logic further in to ConstInfo in a second step (after testing etc etc)

File: compiler/optimizations/inferConstRefs.cpp [hunk 5]
```diff
@@ -23,62 +23,109 @@
...
+  finalizedConstness = q == QUAL_CONST_REF || q == QUAL_CONST_VAL;
+  if (ArgSymbol* arg = toArgSymbol(s)) {
+    // Work around a bug where we can have an arg with the in-intent, but it's
+    // a reference
+    // BHARSH TODO: this shouldn't be possible in the qualified refs impl.
+    if (arg->intent == INTENT_CONST_IN && arg->isRef()) {
+      finalizedConstness = false;
+    }
+  }
+  finalizedRefToConst = sym->hasFlag(FLAG_REF_TO_CONST);
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: Should FLAG_REF_TO_CONST be a Qualifier?

File: compiler/optimizations/inferConstRefs.cpp [hunk 6]
```diff
@@ -103,163 +150,62 @@ static bool canRHSBeConstRef(CallExpr* parent, CallExpr* rhs, SymExpr* use) {
...
+    case PRIM_GET_REAL:
+    case PRIM_GET_IMAG:
+    case PRIM_ADDR_OF:
+    case PRIM_SET_REFERENCE: {
+      // If LHS is a reference and is not a const-ref, the reference in 'rhs'
+      // should not be considered a const-ref either.
+      //
+      // For the get-member primitives, I intend this to be a safe approach
+      // until we know what const-ref means for fields. Basically, if any field
+      // might be modified I do not consider the base object to be const-ref.
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: If that's the case, why disqualify PRIM_GET_MEMBER_VALUE / PRIM_GET_SVEC_MEMBER_VALUE ? Since these do not return a reference. (Or maybe they do in the case of a reference field? Oh, I see, and then you check if LHS is a reference below? I think it'd be worth commenting on this since it's not obvious at a glance.

File: compiler/optimizations/inferConstRefs.cpp [hunk 7]
```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
...
+    return wasRefToConst;
+  }
+
+  bool isFirstCall = false;
+  if (info && info->alreadyCalled == false) {
+    isFirstCall = true;
+    info->alreadyCalled = true;
+  }
+
+  bool retval = true;
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: here, and in other (long) functions, I think it'd be nicer to call it, say 'isRefToConst' than 'retval'.

File: compiler/optimizations/inferConstRefs.cpp [hunk 8]
```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
...
+      if (isFirstCall && info) {
+        INT_ASSERT(info->fnUses == NULL);
+        info->fnUses = fn->firstSymExpr();
+      } else if (info == NULL) {
+        retval = false;
+      }
+      // Need this part to be re-entrant in case of recursive functions
+      while (info && info->fnUses != NULL && retval) {
+        SymExpr* se = info->fnUses;
+        info->fnUses = se ? se->symbolSymExprsNext : NULL;
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: I'm happy to see use of SymbolSymExprs here, but I wonder if we need some functions to abstract uses of symbolSymExprsNext ? Maybe a 'getNext' style group of functions? Could we simply add 'Symbol::nextSymExpr(SymExpr* lastSymExpr)' ?

**Comment by mppf** [comment]
Generally, I am OK with this, although I brought up some specific issues. Besides that, I think the overall structure of the analysis is a little hard for me to follow (with the re-entrancy/firstCall stuff) and so I'd appreciate a comment somewhere just describing the overall structure (and not going into the details of what inferConst & friends decide - just the outline of the analysis and how they might recurse).

File: compiler/optimizations/inferConstRefs.cpp (hunk 3 above)
Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: Not yet, but that could be a good next step for this optimization.

File: compiler/optimizations/inferConstRefs.cpp (hunk 5 above)
Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: That's a good question, but I don't know what the answer is. I guess I'm not entirely sure how much information we plan on representing with the Qualifier.

File: compiler/optimizations/inferConstRefs.cpp [hunk 9]
```diff
@@ -23,62 +23,109 @@
...
+// finds that the Symbol cannot be const.
+//
+// Once we've made a determination either way, the helper function sets the
+// appropriate 'finalized*' field in its ConstInfo instance. If other Symbols
+// then query the constness of this Symbol, we can quickly determine the
+// answer. The helper functions also reset the list of SymExprs to the head, so
+// that later helper functions can use the list re-entrantly.
+//
+// Until we clarify constness more in the language, this pass will not consider
+// a record to be const if any of its fields are modified.
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: doesn't that effectively mean "this pass will not consider a record to be const"?
Or is it const only if it was marked so by the user?

File: compiler/optimizations/inferConstRefs.cpp (hunk 5 above)
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: Yeah. I view it as connected to the const-language-design questions but it's not necessarily so.

File: compiler/optimizations/inferConstRefs.cpp (hunk 9 above)
Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: A record could be const if it only has one "def" with a move/assign, which would be a bit-copy.

File: compiler/optimizations/inferConstRefs.cpp (hunk 8 above)
Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: That would probably be better than what I'm doing. Would you want me to add it in this PR?

Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: Let's put that in a separate PR. It arguably will not need much testing (since it can be "trivial code movement") and we might spend a while nitpicking its name and arguments etc.

Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: But it would help greatly if you can create such a PR.

File: compiler/optimizations/inferConstRefs.cpp [hunk 10]
```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
...
+        if (SymExpr* se = toSymExpr(call->get(2))) {
+          if (se->isRef() && !inferRefToConst(se->symbol())) {
+            retval = false;
+          }
+        }
+        else {
+          CallExpr* RHS = toCallExpr(call->get(2));
+          INT_ASSERT(RHS);
+          if (RHS->isPrimitive(PRIM_ADDR_OF) ||
+              RHS->isPrimitive(PRIM_SET_REFERENCE)) {
```
Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: We're looking for this kind of pattern:

PRIM_MOVE use, <expr>

'use' should always be call->get(1). The branch with addr-of and set-reference will be entered if the RHS of the move is not a SymExpr.

File: compiler/optimizations/inferConstRefs.cpp [hunk 11]
```diff
@@ -350,59 +411,281 @@ static bool inferConstRef(Symbol* sym) {
...
+              }
+            }
+          } else {
+            isRefToConst = false;
+          }
+        }
+      }
+    }
+    else if (call->isResolved()) {
+      isRefToConst = true;
```
Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: I think this case is handled above, right? Maybe say so in a comment?

Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: The case above is a formal looking at its corresponding actuals, because a formal can't be ref-to-const unless all of its actuals are const-to-ref.

When 'sym' is an actual, in the case you've commented on here, we don't care whether or not the formal is ref-to-const.

Comment (mppf) [inline compiler/optimizations/inferConstRefs.cpp]: because if the formal was 'ref', it would be a compiler error?

Comment (benharsh) [inline compiler/optimizations/inferConstRefs.cpp]: Right. 'sym' has to be const-ref to get to this point, or else we'll return early at the top of inferRefToConst. If the formal were a 'ref', then 'sym' could not be a const-ref, meaning that the formal is a const-ref.

Passing a reference actual to a reference formal cannot change what the reference points to, so it should not affect the ref-to-const-ness.

**Comment by benharsh** [comment]
@mppf : @noakesmichael met with me offline and made some good suggestions on how I can clean this code up some more. I'll be working on those changes and will probably not merge this PR today.

**Comment by benharsh** [comment]
@noakesmichael : I've pushed some of the cleanup changes that we discussed. I'll reset my testing status and do more tomorrow (though early gasnet results suggest no failures).

**Comment by benharsh** [comment]
Testing has come back clean.
